### PR TITLE
Hardening

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java [hunk 1]
```diff
@@ -156,9 +155,9 @@ private boolean checkTaskCompleted() throws IOException, InterruptedException {
       log("SonarQube task '%s' status is '%s'", step.taskId, ceTask.getStatus());
       switch (ceTask.getStatus()) {
         case WsClient.CETask.STATUS_SUCCESS:
-          ProjectQualityGate qualityGate = wsClient.getQualityGateWithAnalysisId(ceTask.getAnalysisId());
-          log("SonarQube task '%s' completed. Quality gate is '%s'", step.taskId, qualityGate.getStatus());
-          getContext().onSuccess(new QGStatus(qualityGate.getStatus()));
+          String status = wsClient.requestQualityGateStatus(ceTask.getAnalysisId());
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'requestQualityGateStatus' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I6m_wH-bYHl3c0F&open=AWIv-I6m_wH-bYHl3c0F)

[ISSUE_KEY]: AWIv-I6m_wH-bYHl3c0F

File: src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java [hunk 2]
```diff
@@ -43,38 +44,34 @@ public SQProjectResolver(HttpClient client) {
...
-        // if QG is not available for the project, without errors, projectInfo will be empty
-        return projectInfo;
+      if (version.compareTo(new Version("5.6")) < 0) {
+        Logger.LOG.info(() -> "SQ < 5.6 is not supported");
+        return null;
       }
 
-      getCETask(wsClient, projectInfo, ceTaskId);
+      ProjectInformation projectInfo = new ProjectInformation();
+      projectInfo.setUrl(projectDashboardUrl);
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'setUrl' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I51_wH-bYHl3c0E&open=AWIv-I51_wH-bYHl3c0E)

[ISSUE_KEY]: AWIv-I51_wH-bYHl3c0E

File: src/test/java/hudson/plugins/sonar/utils/SonarUtilsTest.java [hunk 3]
```diff
@@ -45,34 +41,6 @@
   @Rule
   public TemporaryFolder temp = new TemporaryFolder();
```
Comment (SonarTech) [inline src/test/java/hudson/plugins/sonar/utils/SonarUtilsTest.java]: Code Smell Code Smell: Remove this unused "TemporaryFolder". ([squid:S2924](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2924))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I78_wH-bYHl3c0G&open=AWIv-I78_wH-bYHl3c0G)

[ISSUE_KEY]: AWIv-I78_wH-bYHl3c0G

Comment (SonarTech) [inline src/test/java/hudson/plugins/sonar/utils/SonarUtilsTest.java]: Code Smell Code Smell: Remove this unused "TemporaryFolder". ([squid:S2924](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2924))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I78_wH-bYHl3c0G&open=AWIv-I78_wH-bYHl3c0G)

[ISSUE_KEY]: AWIv-I78_wH-bYHl3c0G

File: src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java (hunk 1 above)
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'requestQualityGateStatus' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I6m_wH-bYHl3c0F&open=AWIv-I6m_wH-bYHl3c0F)

[ISSUE_KEY]: AWIv-I6m_wH-bYHl3c0F

File: src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java (hunk 2 above)
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'setUrl' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I51_wH-bYHl3c0E&open=AWIv-I51_wH-bYHl3c0E)

[ISSUE_KEY]: AWIv-I51_wH-bYHl3c0E

Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'setUrl' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I51_wH-bYHl3c0E&open=AWIv-I51_wH-bYHl3c0E)

[ISSUE_KEY]: AWIv-I51_wH-bYHl3c0E

File: src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java (hunk 1 above)
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'requestQualityGateStatus' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I6m_wH-bYHl3c0F&open=AWIv-I6m_wH-bYHl3c0F)

[ISSUE_KEY]: AWIv-I6m_wH-bYHl3c0F

File: src/test/java/hudson/plugins/sonar/utils/SonarUtilsTest.java (hunk 3 above)
Comment (SonarTech) [inline src/test/java/hudson/plugins/sonar/utils/SonarUtilsTest.java]: Code Smell Code Smell: Remove this unused "TemporaryFolder". ([squid:S2924](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2924))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIv-I78_wH-bYHl3c0G&open=AWIv-I78_wH-bYHl3c0G)

[ISSUE_KEY]: AWIv-I78_wH-bYHl3c0G

File: src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java [hunk 4]
```diff
@@ -43,38 +41,35 @@ public SQProjectResolver(HttpClient client) {
...
   @CheckForNull
-  public ProjectInformation resolve(@Nullable String projectUrl, @Nullable String ceTaskId, String installationName) {
+  public ProjectInformation resolve(@Nullable String serverUrl, @Nullable String projectDashboardUrl, String ceTaskId, String installationName) {
     SonarInstallation inst = SonarInstallation.get(installationName);
     if (inst == null) {
       Logger.LOG.info(() -> "Invalid installation name: " + installationName);
       return null;
     }
+    if (serverUrl == null) {
+      Logger.LOG.info(() -> String.format(Locale.US, "No server url."));
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java]: Code Smell Code Smell: String contains no format specifiers. ([squid:S3457](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3457))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIwh_yk_wH-bYHl3dHv&open=AWIwh_yk_wH-bYHl3dHv)

[ISSUE_KEY]: AWIwh_yk_wH-bYHl3dHv

File: src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java [hunk 5]
```diff
@@ -127,6 +135,7 @@ private void processStepParameters() throws IOException, InterruptedException {
       for (SonarAnalysisAction a : actions) {
         if (a.getCeTaskId() != null) {
           step.setTaskId(a.getCeTaskId());
+          step.setServerUrl(a.getServerUrl());
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'setServerUrl' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIwh_zS_wH-bYHl3dHw&open=AWIwh_zS_wH-bYHl3dHw)

[ISSUE_KEY]: AWIwh_zS_wH-bYHl3dHw

File: src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java [hunk 6]
```diff
@@ -127,6 +142,7 @@ private void processStepParameters() throws IOException, InterruptedException {
...
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'setServerUrl' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIwh_zS_wH-bYHl3dHw&open=AWIwh_zS_wH-bYHl3dHw)

[ISSUE_KEY]: AWIwh_zS_wH-bYHl3dHw

File: src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java [hunk 7]
```diff
@@ -196,7 +220,7 @@ public void onTaskCompleted(String taskId, String taskStatus, @Nullable String q
         SonarQubeWebHook.get().removeListener(this);
         switch (taskStatus) {
           case WsClient.CETask.STATUS_SUCCESS:
-            getContext().onSuccess(new QGStatus(qgStatus));
+            handleQGStatus(qgStatus);
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'handleQGStatus' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWI9oBgHdpSr17abUG11&open=AWI9oBgHdpSr17abUG11)

[ISSUE_KEY]: AWI9oBgHdpSr17abUG11

File: src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java (hunk 4 above)
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/client/SQProjectResolver.java]: Code Smell Code Smell: String contains no format specifiers. ([squid:S3457](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3457))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWIwh_yk_wH-bYHl3dHv&open=AWIwh_yk_wH-bYHl3dHv)

[ISSUE_KEY]: AWIwh_yk_wH-bYHl3dHv

File: src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java [hunk 8]
```diff
@@ -196,7 +227,7 @@ public void onTaskCompleted(String taskId, String taskStatus, @Nullable String q
...
```
Comment (SonarTech) [inline src/main/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStep.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'handleQGStatus' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWI9oBgHdpSr17abUG11&open=AWI9oBgHdpSr17abUG11)

[ISSUE_KEY]: AWI9oBgHdpSr17abUG11

File: src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java [hunk 9]
```diff
@@ -71,56 +70,52 @@ public void setUp(Context context, Run<?, ?> build, FilePath workspace, Launcher
...
-    map.put("SONAR_JDBC_URL", getOrDefault(inst.getDatabaseUrl(), ""));
-
-    String jdbcDefault = SQServerVersions.SQ_5_1_OR_LOWER.equals(inst.getServerVersion()) ? DEFAULT_SONAR : "";
-    map.put("SONAR_JDBC_USERNAME", getOrDefault(inst.getDatabaseLogin(), jdbcDefault));
-    map.put("SONAR_JDBC_PASSWORD", getOrDefault(inst.getDatabasePassword(), jdbcDefault));
-
     if (StringUtils.isEmpty(inst.getMojoVersion())) {
       map.put("SONAR_MAVEN_GOAL", "sonar:sonar");
     } else {
       map.put("SONAR_MAVEN_GOAL", SonarUtils.getMavenGoal(inst.getMojoVersion()));
```
Comment (SonarTech) [inline src/main/java/hudson/plugins/sonar/SonarBuildWrapper.java]: Code Smell Code Smell: Annotate the parameter with @javax.annotation.Nullable in method 'getMavenGoal' declaration, or make sure that null can not be passed as argument. ([squid:S4449](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS4449))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.jenkins-ci.plugins%3Asonar&pullRequest=90&issues=AWI-fL1adpSr17abUILf&open=AWI-fL1adpSr17abUILf)

[ISSUE_KEY]: AWI-fL1adpSr17abUILf

File: src/main/java/hudson/plugins/sonar/client/WsClient.java [hunk 10]
```diff
@@ -19,52 +19,36 @@
...
   public static final String API_PROJECT_STATUS_WITH_ANALYSISID = "/api/qualitygates/project_status?analysisId=";
   public static final String API_VERSION = "/api/server/version";
   public static final String API_PROJECT_NAME = "/api/projects/index?format=json&key=";
   public static final String API_CE_TASK = "/api/ce/task?id=";
 
   private final HttpClient client;
   private final String serverUrl;
-  private final String username;
-  private final String password;
+  private final String token;
```
[the remark this answers is not in the document]
Comment (henryju) [reply src/main/java/hudson/plugins/sonar/client/WsClient.java]: AFAIK, @CheckForNull should only be used on getters. Inside a class, the analyzer is able to guess that the attribute is nullable.

**Pull request merged**